engine/engine.go

```diff
@@ -584,3 +566,27 @@ func bk2progrock(event *bkclient.SolveStatus) *progrock.StatusUpdate {
 
 	return &status
 }
+
+func progrockForwarder(w progrock.Writer) (string, progrock.Writer, func() error, error) {
+	progSock := filepath.Join(
+		xdg.CacheHome,
+		"dagger",
+		fmt.Sprintf("progrock-%d.sock", time.Now().UnixNano()),
+	)
+
+	if err := os.MkdirAll(filepath.Dir(progSock), 0700); err != nil {
+		return "", nil, nil, err
+	}
+
+	l, err := net.Listen("unix", progSock)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	progW, err := progrock.ServeRPC(l, w)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return progSock, progW, l.Close, nil
```

Comment:
nit: unix sock won't be deleted from the fs when it's closed, so we could toss in an `os.Remove` with the cleanup too. Otherwise I suppose over time you will keep accumulating leftover sockets indefinitely, which is mostly harmless but technically takes up inodes I suppose.

Reply:
Go stdlib will unlink on close for Unix sockets:

https://github.com/golang/go/blob/72f448cb48069ddeb0c30e726fc3b9d113aaf51e/src/net/unixsock_posix.go#L193

I confirmed they're being cleaned up locally, but it's definitely possible to leak them over time if you aggressively kill the `dagger` process, and I had one leaked probably from a `SIGQUIT` I sent earlier. Maybe this could be switched to TCP? :thinking:

Reply:
> Go stdlib will unlink on close for Unix sockets:

Oh cool, I had no idea, that's very nice of it!

> I confirmed they're being cleaned up locally, but it's definitely possible to leak them over time if you aggressively kill the dagger process, and I had one leaked probably from a SIGQUIT I sent earlier. Maybe this could be switched to TCP? 🤔

Eh I wouldn't worry about it then. I was just concerned that like 1000 runs of dagger would result in 1000 sockets, but if it only rarely leaks it really doesn't matter in practice I don't think.

Reply:
:+1: sounds good! I think putting it in `xdg.RuntimeDir` should help too, since that tends to be a directory that gets blown away when the machine reboots, so any leaks won't last for long.